experimental/dds/tree2/src/core/tree/visitDelta.ts

```diff
@@ -3,319 +3,681 @@
+	/**
+	 * Replaces a range of nodes in the current field by transferring them out to a new detached field
+	 * and transferring in all the nodes from an existing detached field in their place.
+	 * The number of nodes being detached must match the number of nodes being attached.
+	 * @param newContentSource - The detached field to transfer the new nodes from.
+	 * @param range - The bounds of the range of nodes to replace.
+	 * @param oldContentDestination - The key for a new detached field to transfer the old nodes to.
+	 * @param kind - The kinds of replacement this operation conveys.
+	 */
+	replace(
+		newContentSource: FieldKey,
+		range: Range,
+		oldContentDestination: FieldKey,
+		kind: ReplaceKind,
+	): void;
+
+	enterNode(index: NodeIndex): void;
+	exitNode(index: NodeIndex): void;
 	enterField(key: FieldKey): void;
 	exitField(key: FieldKey): void;
 }
 
 interface PassConfig {
+	readonly name: "detach" | "attach";
 	readonly func: Pass;
-	readonly applyValueChanges: boolean;
+	readonly treeIndex: TreeIndex;
 
-	readonly modsToMovedTrees: Map<Delta.MoveId, Delta.HasModifications>;
-	readonly movedOutRanges: RangeMap<Delta.MoveId>;
+	/**
+	 * The new trees that need to be created.
+	 * Only used in the detach pass.
+	 */
+	readonly creations: Map<Delta.Insert, ForestRootId>;
+	/**
+	 * Nested changes that need to be applied to detached roots.
+	 * In the detach pass, this is used for:
+	 * - Changes under newly created trees (since they are created in a detached field)
+	 * - Changes under existing trees that start out as being removed (no matter what happens to them during this visit)
+	 * In the attach pass, this is used for:
+	 * - Changes under trees end up as removed as part of this visit (no matter what happened to them during this visit)
+	 */
+	readonly rootChanges: Map<ForestRootId, Delta.FieldMarks>;
+	/**
+	 * Represents transfers of roots from one detached field to another.
+	 * In the detach pass, this is used for:
+	 * - Transferring a removed node (that is being moved) to the detached field that corresponds to the move ID
+	 * In the attach pass, this is used for:
+	 * - Transferring a created node (that is transient) to the detached field that corresponds to its detach ID
+	 * - Transferring a restored node (that is transient) to the detached field that corresponds to its detach ID
+	 * - Transferring a moved node (that is removed) to the detached field that corresponds to its detach ID
+	 * TODO#5481: update the TreeIndex instead of moving the nodes around.
+	 */
+	readonly rootToRootTransfers: RootTransfers;
+	/**
+	 * Stores the nested changes for all moved trees.
+	 */
+	readonly modsToMovedTrees: Map<Delta.MoveId, Delta.FieldMarks>;
+	/**
+	 * All creations are made in a detached field and eventually transferred to their final destination.
+	 * This map keeps track of the root ID where creations occur.
+	 */
+	readonly insertToRootId: Map<Delta.Insert, ForestRootId>;
```

Comment:
There doesn't seem to be any difference between this and creations, maybe remove one?